### PR TITLE
101390552 Event manager can request remit

### PR DESCRIPTION
**Pull request by andela-oesho**
#### What does this PR do?

Allows an event manager to request remit 5 days after an event has ended

#### Description of Task to be completed?
Validate events that can request remit
Send email to an event manager if remittance was successful

#### How should this be manually tested?
Login into eventx as an event manager and visit the eventlist dashboard to request remittance, if successful a mail will be recieved.

#### Screenshots (if appropriate)
image

#### Questions:
Are there better ways to complete this feature?

### CODE REVIEW

File: app/assets/stylesheets/remit.scss [hunk 1]
```diff
@@ -0,0 +1,4 @@
+// Place all the styles related to the remit controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
+
```
Comment (andela-iarisah) [inline app/assets/stylesheets/remit.scss]: since this file is not being used, you can delete :)

File: app/controllers/remit_controller.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class RemitController < ApplicationController
+  def new
+    @event = Event.find_by(id: params[:id])
```
Comment (andela-iarisah) [inline app/controllers/remit_controller.rb]: Instead of using `find_by`, you can use `Event.find(params[:event_id])`

File: app/decorators/remit_decorator.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+  # Define presentation-specific methods here. Helpers are accessed through
+  # `helpers` (aka `h`). You can override attributes, for example:
+  #
+  #   def created_at
+  #     helpers.content_tag :span, class: 'time' do
+  #       object.created_at.strftime("%a %m/%d/%y")
+  #     end
+  #   end
+end
```
Comment (andela-iarisah) [inline app/decorators/remit_decorator.rb]: same as for the remit css file

File: app/controllers/remit_controller.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    @manager = @event.manager_profile
+    @remit = @event.build_remit
+    if @remit.save
+      @remit
+    else
+      flash[:notice] = "Event cannot yet request a remittance!"
+      redirect_to my_events_path
```
Comment (andela-iarisah) [inline app/controllers/remit_controller.rb]: `redirect_to my_events_path, notice: "Event cannot yet requ......`

 a one-liner

File: app/helpers/remit_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module RemitHelper
+end
```
Comment (andela-iarisah) [inline app/helpers/remit_helper.rb]: same as for the decorator

File: app/mailers/remit_mailer.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class RemitMailer < ApplicationMailer
+  def remitance_confirmation(event, manager, remit)
+    @event = event
+    @manager = manager
+
+    @remit = remit
+    mail(to: manager.company_mail, subject: "Remittance Ticket Fee")
+  end
+end
```
Comment (andela-iarisah) [inline app/mailers/remit_mailer.rb]: 👍

File: app/models/event.rb [hunk 7]
```diff
@@ -108,6 +106,10 @@ def self.find_event(params)
     end
   end
 
+  def can_request_remit?
+    Date.today >= (end_date + 5.days)
+  end
+
```
Comment (andela-iarisah) [inline app/models/event.rb]: 👍

File: app/models/remit.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  before_save :set_total_amount
+  after_create :send_remit_mail
+  validate :check_remit_status
+
+  def check_remit_status
+    if event.can_request_remit?
+      true
+    else
+      errors.add(:event, "Event cannot yet request a remittance!")
+    end
```
Comment (andela-iarisah) [inline app/models/remit.rb]: 👍

File: config/environments/development.rb [hunk 9]
```diff
@@ -23,17 +23,10 @@
...
-  config.action_mailer.smtp_settings = {
-    address: "smtp.gmail.com",
-    port: 587,
-    domain: "gmail.com",
-    authentication: "plain",
-    #   openssl_verify_mode:'none' ,
-    enable_starttls_auto: true,
-    user_name: ENV["GMAIL_USERNAME"],
-    password: ENV["GMAIL_PASSWORD"]
-  }
```
Comment (andela-iarisah) [inline config/environments/development.rb]: 👍

File: test/controllers/remit_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class RemitControllerTest < ActionController::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
Comment (andela-iarisah) [inline test/controllers/remit_controller_test.rb]: delete test directory

File: spec/controllers/remit_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    it "does not get new remit when remit is requested for" do
+      @event.update(start_date: Time.now - 10.days)
+      @event.update(end_date: Time.now - 3.days)
+
+      get :new, id: @event.id
+
+      expect(flash[:notice]).to eq "Event cannot yet request a remittance!"
+    end
+  end
+end
```
Comment (andela-iarisah) [inline spec/controllers/remit_controller_spec.rb]: Just to be on the safe side, you can add a scenario for when the manager has already asked for remittance.

In this case, there should be a message telling him that he has already been remitted.

File: app/assets/stylesheets/remit.scss (hunk 1 above)
Comment (andela-oesho) [inline app/assets/stylesheets/remit.scss]: thanks noted

File: app/controllers/remit_controller.rb (hunk 4 above)
Comment (andela-oesho) [inline app/controllers/remit_controller.rb]: Rubocop/Hound  will complain of long lines if i implement it in one line

File: test/controllers/remit_controller_test.rb (hunk 10 above)
Comment (andela-oesho) [inline test/controllers/remit_controller_test.rb]: thanks noted

File: spec/controllers/remit_controller_spec.rb (hunk 11 above)
Comment (andela-oesho) [inline spec/controllers/remit_controller_spec.rb]: Thank you for the suggestion, it will be implemented in the next PR

File: app/controllers/remit_controller.rb (hunk 2 above)
Comment (andela-oesho) [inline app/controllers/remit_controller.rb]: i used find by because find will return an error if not found, whereas find_by will return null.

File: spec/controllers/remit_controller_spec.rb (hunk 11 above)
Comment (andela-sdamian) [inline spec/controllers/remit_controller_spec.rb]: @andela-iarisah I thought scenarios are terms used for `feature`  test not controller test. 😕

Comment (andela-iarisah) [inline spec/controllers/remit_controller_spec.rb]: @andela-oesho, like I said earlier, this is still part of this story. You have to try to cover edge cases when working on features. This is not a story creep or a scope creep. It's a scenario in the story not covered. Add it to this PR.
We'll probably need to do a more thorough story check next time.

@andela-sdamian, scenarios are keywords used in feature test but it's still an English word used to define how an event occurs :)

File: app/controllers/remit_controller.rb (hunk 2 above)
Comment (andela-iarisah) [inline app/controllers/remit_controller.rb]: Ah I see. It's because of the routes. If a manager gets remitted from his events, then the routes are wrong. I'm not too sure how it works any more so perhaps we can meet today?

File: app/controllers/remit_controller.rb (hunk 4 above)
Comment (andela-iarisah) [inline app/controllers/remit_controller.rb]: it's not just the one-liner part. It's also the current way of writing this. I think this convention is Rails 3. If Rubocop bugs you too much, you can increase the length allowed to `80` 
OR 
break it
```
redirect_to my_events_path,
notice: "Event.....
```

File: spec/controllers/remit_controller_spec.rb (hunk 11 above)
Comment (andela-oesho) [inline spec/controllers/remit_controller_spec.rb]: @andela-iarisah i think my reply was misread i meant it will be handled in the next updated PR for this same story which has currently been handled and should be ready for review on Monday.

Comment (andela-sdamian) [inline spec/controllers/remit_controller_spec.rb]: @andela-iarisah O yeah, I see the ambiguity.
